### DISCUSSION TITLE
大表联表查询造成be崩溃

### DISCUSSION THREAD

**Discussion by gaojunjie03** [apache/doris#14631] [A - General / Q&A]
1000w表对2000w表联表查询，be崩溃，错误信息：
`E1128 11:40:37.278808 48630 allocator.h:142] Allocator: Cannot mmap 68719476736.
W1128 11:40:37.279067 48630 mem_tracker_limiter.cpp:204] Allocator: Cannot mmap 68719476736.
    process memory used 2.10 GB, limit 37.89 GB, hard limit 42.63 GB, tc/jemalloc allocator cache 65.38 MB
    MemTrackerLimiter Label=, Type=consistency, Limit=-1.00 B(-1 B), Used=0(0 B), Peak=0(0 B)
    MemTrackerLimiter Label=, Type=batch_load, Limit=-1.00 B(-1 B), Used=0(0 B), Peak=0(0 B)
    MemTrackerLimiter Label=, Type=clone, Limit=-1.00 B(-1 B), Used=0(0 B), Peak=0(0 B)
    MemTrackerLimiter Label=, Type=schema_change, Limit=-1.00 B(-1 B), Used=0(0 B), Peak=0(0 B)
    MemTrackerLimiter Label=, Type=compaction, Limit=-1.00 B(-1 B), Used=0(0 B), Peak=0(0 B)
    MemTrackerLimiter Label=, Type=load, Limit=-1.00 B(-1 B), Used=0(0 B), Peak=0(0 B)
    MemTrackerLimiter Label=, Type=query, Limit=-1.00 B(-1 B), Used=-62.78 GB(-67410451131 B), Peak=999.28 MB(1047820313 B)
    MemTrackerLimiter Label=, Type=global, Limit=-1.00 B(-1 B), Used=101.03 GB(108475053759 B), Peak=101.03 GB(108475053759 B)
    MemTrackerLimiter Label=, Type=tc/jemalloc_cache, Limit=-1.00 B(-1 B), Used=65.38 MB(68557896 B), Peak=-1.00 B(-1 B)
    MemTrackerLimiter Label=, Type=process, Limit=-1.00 B(-1 B), Used=38.31 GB(41133160524 B), Peak=-1.00 B(-1 B)
    MemTrackerLimiter Label=Orphan, Type=global, Limit=-1.00 B(-1 B), Used=100.36 GB(107762587490 B), Peak=36.36 GB(39043110785 B)
    MemTracker Label=BufferAllocator, Parent Label=Orphan, Used=0(0 B), Peak=0(0 B)
    MemTracker Label=LoadChannelMgr, Parent Label=Orphan, Used=0(0 B), Peak=0(0 B)
    MemTracker Label=StorageEngine, Parent Label=Orphan, Used=2.82 MB(2958424 B), Peak=2.32 MB(2429152 B)
    MemTracker Label=SegCompaction, Parent Label=Orphan, Used=0(0 B), Peak=0(0 B)
    MemTracker Label=SegmentMeta, Parent Label=Orphan, Used=287.07 KB(293963 B), Peak=287.05 KB(293943 B)
    MemTracker Label=TabletManager, Parent Label=Orphan, Used=0(0 B), Peak=0(0 B)
    MemTrackerLimiter Label=DataPageCache, Type=global, Limit=-1.00 B(-1 B), Used=299.71 MB(314264815 B), Peak=298.67 MB(313179217 B)
    MemTrackerLimiter Label=IndexPageCache, Type=global, Limit=-1.00 B(-1 B), Used=48.93 MB(51311214 B), Peak=47.91 MB(50240353 B)
    MemTrackerLimiter Label=SegmentCache, Type=global, Limit=-1.00 B(-1 B), Used=0(0 B), Peak=0(0 B)
    MemTrackerLimiter Label=DiskIO, Type=global, Limit=37.89 GB(40685115801 B), Used=0(0 B), Peak=0(0 B)
    MemTrackerLimiter Label=ChunkAllocator, Type=global, Limit=-1.00 B(-1 B), Used=330.82 MB(346890240 B), Peak=421.20 MB(441655296 B)
    MemTrackerLimiter Label=LastestSuccessChannelCache, Type=global, Limit=-1.00 B(-1 B), Used=0(0 B), Peak=0(0 B)
    MemTrackerLimiter Label=DeleteBitmap AggCache, Type=global, Limit=-1.00 B(-1 B), Used=0(0 B), Peak=0(0 B)
*** Query id: 8cfce1e528434fa2-8e26c0bf1d66831b ***
*** Aborted at 1669606837 (unix time) try "date -d @1669606837" if you are using GNU date ***
*** Current BE git commitID: 24b51b9 ***
*** SIGSEGV address not mapped to object (@0xf) received by PID 48408 (TID 0xfffc38f88340) from PID 15; stack trace: ***
 0# doris::signal::(anonymous namespace)::FailureSignalHandler(int, siginfo_t*, void*) at /opt/doris/be/src/common/signal_handler.h:420
 1# os::Linux::chained_handler(int, siginfo_t*, void*) in /opt/jdk1.8.0_311/jre/lib/aarch64/server/libjvm.so

bin/start_be.sh：行 235: 48408 段错误               （核心已转储）${LIMIT:+${LIMIT}} "${DORIS_HOME}/lib/doris_be" "$@" 2>&1 < /dev/null`

随后我用1000w-1000w联表查询是可以得，

**Comment by LemonLiTree** [comment]
doris 哪个版本？

**Reply by gaojunjie03** [reply]
master 分支  
commitID: 24b51b9

**Comment by LemonLiTree** [comment]


**Reply by gaojunjie03** [reply]
我学习调研后发现，是因为我写的join sql很粗略，纯粹只是测试doris的相关性能。
这个sql的join key并没有用当时创建表的分通列进行创建，从而在doris执行sql时是partitioned join shuffle join等的方式进行join查询，数据量过大造成内存不足而崩溃，后面我学习了下doris的相关join优化原理，使用分桶列进行的join查询，explain出来的是bucket shuffle join形式，这个时候不会报错也不会崩溃，sql能正常执行并返回结果。

不过给点建议就是，即使sql写的很粗也尽可能不要把be搞崩溃吧，毕竟在搭建数仓后，对于外部人员（只管写sql进行分析）来说，他们可能不清楚这些doris的机制，随便创建一个表 没有做好分桶分区等，然后写一条sql就会造成崩溃了的。

**Reply by LemonLiTree** [reply]
exec_mem_limit有设置吗？

**Reply by gaojunjie03** [reply]
> exec_mem_limit有设置吗？
设置了  设置成8G 10G都会造成崩溃

**Reply by LemonLiTree** [reply]
方便发一哈建表和sql吗，可以脱敏下

**Reply by gaojunjie03** [reply]
CREATE TABLE t_car_test (
cardid varchar(30) ,
plate varchar(30) ,
caacity varchar(30) ,
rult varchar(30) ,
coltime varchar(30) ,
fidence varchar(30) ,
tent varchar(30) ,
deid varchar(30) ,
dtype varchar(30) ,
dent varchar(30) ,
eood varchar(30) ,i
eupci varchar(30) ,
laber varchar(30) ,
lon varchar(30) ,
plpe varchar(30) ,
poting varchar(30) ,
reer varchar(30) ,
ip varchar(30)
)
UNIQUE KEY(cardid)
DISTRIBUTED BY HASH(cardid) BUCKETS 10

cardid分桶

但下面得sql没有用cardid进行join key
select a.plate,a.plpe,a.cardid,b.plate,b.plpe,b.cardid from t_car_test_1000w a left join t_car_test_2000w b on a.fidence=b.caacity limit 100

数据可以随便造一点字符串进去就可以了，根据ddl语句创建两个表 一个存1000w数据一个存2000w数据,
如以下数据
793207439003820032 | 粤ECWKSD | 0 | BCC | 2022-11-24 17:24:16 | 0 |   | 793207439003820033 | FIRST_FOUND | Y5ITQ | 793207439003820034 | 793207439003820035 | 3 |   | 3 | JUPVU | 58YGKAHQDE | 11.12.10.57

**Reply by LemonLiTree** [reply]
好的，感谢

**Reply by gaojunjie03** [reply]
> 好的，感谢

你好，请问下这个有解决方案了么？ 我用最新的1.2.0-rc04 版本也会崩溃，当数据量大，且不用分桶列进行join查询的时候就会崩溃

**Comment by hf200012** [comment]
@gaojunjie03 机器什么配置

**Reply by gaojunjie03** [reply]
> @gaojunjie03 机器什么配置

3台机器，每台机器都装有fe和be， 每台机器都是32G内存 arm64
